qiskit/algorithms/phase_estimators/iphase_estimation.py

```diff
@@ -0,0 +1,193 @@
+            qc.append(problem['state_preparation'], eigenstate_register)
+        elif problem['state_preparation'] is not None:
+            qc += problem['state_preparation'].construct_circuit('circuit', eigenstate_register)
+        # hadamard on phase_register[0]
+        qc.h(phase_register[0])
+        # controlled-U
+        # TODO: We may want to allow flexibility in how the power is computed
+        # For example, it may be desirable to compute the power via Trotterization, if
+        # we are doing Trotterization anyway.
+        unitary_power = problem['unitary'].power(2 ** (k - 1)).control()
+        qc.append(unitary_power, list(range(1, problem['unitary'].num_qubits + 1)) + [0])
+        qc.p(omega, phase_register[0])
+        # hadamard on phase_register[0]
+        qc.h(phase_register[0])
+        if measurement:
+            c = ClassicalRegister(1, name='c')
+            qc.add_register(c)
+            qc.measure(phase_register, c)
+        return qc
+
+    def _estimate_phase_iteratively(self, problem):
+        """
+        Main loop of iterative phase estimation.
+        """
+        omega_coef = 0
+        # k runs from the number of iterations back to 1
+        for k in range(problem['num_iterations'], 0, -1):
+            omega_coef /= 2
+            if self._quantum_instance.is_statevector:
+                qc = self.construct_circuit(problem, k, -2 * numpy.pi * omega_coef,
+                                            measurement=False)
+                result = self._quantum_instance.execute(qc)
+                complete_state_vec = result.get_statevector(qc)
+                ancilla_density_mat = qiskit.quantum_info.partial_trace(
+                    complete_state_vec,
+                    range(problem['unitary'].num_qubits)
+                )
+                ancilla_density_mat_diag = numpy.diag(ancilla_density_mat)
+                max_amplitude = max(ancilla_density_mat_diag.min(),
+                                    ancilla_density_mat_diag.max(), key=abs)
+                x = numpy.where(ancilla_density_mat_diag == max_amplitude)[0][0]
+            else:
+                qc = self.construct_circuit(problem, k, -2 * numpy.pi * omega_coef,
+                                            measurement=True)
+                measurements = self._quantum_instance.execute(qc).get_counts(qc)
+
+                if '0' not in measurements:
+                    if '1' in measurements:
+                        x = 1
+                    else:
+                        raise RuntimeError('Unexpected measurement {}.'.format(measurements))
+                else:
+                    if '1' not in measurements:
+                        x = 0
+                    else:
+                        x = 1 if measurements['1'] > measurements['0'] else 0
+            omega_coef = omega_coef + x / 2
+        return omega_coef
+
+    def estimate(self,
+                 num_iterations: int,
```

Comment:
Should we add a check that the iterations are larger than 0? (Plus a test checking that)

Reply:
Done